src/serializers/type_serializers/union.rs

```diff
@@ -198,24 +202,180 @@ impl TypeSerializer for UnionSerializer {
     }
 }
 
-pub struct TaggedUnionBuilder;
+#[derive(Debug, Clone)]
+pub struct TaggedUnionSerializer {
+    discriminator: Discriminator,
+    lookup: HashMap<String, CombinedSerializer>,
+    choices: Vec<CombinedSerializer>,
+    name: String,
+}
 
-impl BuildSerializer for TaggedUnionBuilder {
+impl BuildSerializer for TaggedUnionSerializer {
     const EXPECTED_TYPE: &'static str = "tagged-union";
 
     fn build(
         schema: &Bound<'_, PyDict>,
         config: Option<&Bound<'_, PyDict>>,
         definitions: &mut DefinitionsBuilder<CombinedSerializer>,
     ) -> PyResult<CombinedSerializer> {
-        let schema_choices: Bound<'_, PyDict> = schema.get_as_req(intern!(schema.py(), "choices"))?;
-        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(schema_choices.len());
+        let py = schema.py();
+        let discriminator = Discriminator::new(py, &schema.get_as_req(intern!(py, "discriminator"))?)?;
+
+        let choice_list: Bound<PyDict> = schema.get_as_req(intern!(py, "choices"))?;
+        let mut lookup: HashMap<String, CombinedSerializer> = HashMap::with_capacity(choice_list.len());
+        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(choice_list.len());
+
+        for (choice_key, choice_schema) in choice_list {
+            let serializer = CombinedSerializer::build(choice_schema.downcast()?, config, definitions).unwrap();
+            choices.push(serializer.clone());
+            lookup.insert(choice_key.to_string(), serializer);
+        }
+
+        let descr = choices
+            .iter()
+            .map(TypeSerializer::get_name)
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        Ok(Self {
+            discriminator,
+            lookup,
+            choices,
+            name: format!("TaggedUnion[{descr}]"),
+        }
+        .into())
+    }
+}
+
+impl_py_gc_traverse!(TaggedUnionSerializer { discriminator, lookup });
+
+impl TypeSerializer for TaggedUnionSerializer {
+    fn to_python(
+        &self,
+        value: &Bound<'_, PyAny>,
+        include: Option<&Bound<'_, PyAny>>,
+        exclude: Option<&Bound<'_, PyAny>>,
+        extra: &Extra,
+    ) -> PyResult<PyObject> {
+        let py = value.py();
+
+        let mut new_extra = extra.clone();
+        new_extra.check = SerCheck::Strict;
+
+        if let Some(tag) = self.get_discriminator_value(value) {
```

Comment:
Should we warn if fetching the tag failed?

Reply:
Good idea